File: sdk/ts/src/router/bert.ts
/**
 * BERTRouter — BERT-based sequence classification router.
 *
 * Tokenizes the prompt, sends token IDs and attention mask to a
 * Triton-hosted BERT model, applies softmax to logits, and returns
 * the strong-model win rate.
 *
 * Port of Python's BERTRouter from routellm/routers/routers.py.
 */

import type { Router } from "./router";
import type { InferRequest, InferResponse } from "../triton/client";

// ─── Tokenizer interface (stubbed) ──────────────────────────────────────────

/** Tokenizer output — mirrors HuggingFace tokenizer output shape. */
export interface TokenizerOutput {
  inputIds: number[];
  attentionMask: number[];
}

/**
 * Tokenizer interface for BERT.
 *
 * Implementations should handle padding, truncation, and special tokens
 * (CLS, SEP, etc.) internally.
 */
export interface Tokenizer {
  encode(text: string): TokenizerOutput;
}

// ─── Triton caller interface ─────────────────────────────────────────────────

/**
 * Sends raw inference requests to Triton.
 * Separated from TritonClient so tests can inject a mock.
 */
export interface TritonInfer {
  infer(request: InferRequest): Promise<InferResponse>;
}

// ─── Math helpers ────────────────────────────────────────────────────────────

/**
 * Softmax over an array of logits.
 *
 * Subtracts max for numerical stability, then normalizes.
 */
export function softmax(logits: number[]): number[] {
  const maxVal = Math.max(...logits);
  const exps = logits.map((v) => Math.exp(v - maxVal));
  const sum = exps.reduce((a, b) => a + b, 0);
  return exps.map((e) => e / sum);
}

// ─── Options ─────────────────────────────────────────────────────────────────

export interface BERTRouterOptions {
  /** Triton inference caller for the BERT model. */
  triton: TritonInfer;
  /** Tokenizer for the BERT model. */
  tokenizer: Tokenizer;
  /** Number of output labels (default: 3). */
  numLabels?: number;
  /** Name for the input_ids tensor (default: "input_ids"). */
  inputIdsName?: string;
  /** Name for the attention_mask tensor (default: "attention_mask"). */
  attentionMaskName?: string;
  /** Name for the output logits tensor (default: "logits"). */
  outputName?: string;
}

// ─── Router ──────────────────────────────────────────────────────────────────

export class BERTRouter implements Router {
  private readonly _triton: TritonInfer;
  private readonly _tokenizer: Tokenizer;
  private readonly _numLabels: number;
  private readonly _inputIdsName: string;
  private readonly _attentionMaskName: string;
  private readonly _outputName: string;

  constructor(opts: BERTRouterOptions) {
    this._triton = opts.triton;
    this._tokenizer = opts.tokenizer;
    this._numLabels = opts.numLabels ?? 3;
    this._inputIdsName = opts.inputIdsName ?? "input_ids";
    this._attentionMaskName = opts.attentionMaskName ?? "attention_mask";
    this._outputName = opts.outputName ?? "logits";
  }

  /**
   * Score a prompt using BERT classification.
   *
   * 1. Tokenize the prompt.
   * 2. Send input_ids + attention_mask to Triton.
   * 3. Apply softmax to logits.
   * 4. Compute strong-model win rate as 1 - P(tie or weak wins).
   *
   * The Python code uses the last two softmax classes as the
   * "binary probability" of tie + weak winning.
   */
  async score(prompt: string): Promise<number> {
    const { inputIds, attentionMask } = this._tokenizer.encode(prompt);

    const request: InferRequest = {
      inputs: [
        {
          name: this._inputIdsName,
          shape: [1, inputIds.length],
          datatype: "INT64",
          data: inputIds,
        },
        {
          name: this._attentionMaskName,
          shape: [1, attentionMask.length],
          datatype: "INT64",
          data: attentionMask,
        },
      ],
      outputs: [{ name: this._outputName }],
    };

    const response = await this._triton.infer(request);
    const logitsOutput = response.outputs.find(
      (o) => o.name === this._outputName,
    );

    if (!logitsOutput || !logitsOutput.data.length) {
      // Default to strong model on error.
      return 1;
    }

    const logits = logitsOutput.data.slice(0, this._numLabels) as number[];
    const probs = softmax(logits);

    // binaryProb = sum of last 2 classes (tie + weak wins).
    const binaryProb =
      probs.length >= 2
        ? probs[probs.length - 2] + probs[probs.length - 1]
        : 0;

    return 1 - binaryProb;
  }
}

File: go/ai/llm/router/bert.go
package router

import (
	"context"
	"fmt"
	"math"

	"hop.top/kit/go/ai/llm/triton"
)

// Tokenizer converts text into token IDs for BERT-style models.
// Implementations should handle padding and truncation.
type Tokenizer interface {
	// Tokenize returns input IDs for the given text.
	Tokenize(text string) ([]int32, error)
}

// BERTRouter implements BERT-based classification routing.
//
// It tokenizes the prompt locally, sends token IDs to a Triton Scorer,
// and applies softmax locally to compute the strong model's win rate.
// The model outputs 3 logits: [strong_wins, tie, weak_wins].
type BERTRouter struct {
	tokenizer Tokenizer
	scorer    triton.Scorer
	numLabels int
}

// BERTConfig holds configuration for creating a BERTRouter.
type BERTConfig struct {
	Tokenizer Tokenizer
	Scorer    triton.Scorer
	NumLabels int // default: 3
}

// NewBERTRouter creates a BERT classification router.
func NewBERTRouter(cfg BERTConfig) (*BERTRouter, error) {
	if cfg.Tokenizer == nil {
		return nil, fmt.Errorf("bert: tokenizer is required")
	}
	if cfg.Scorer == nil {
		return nil, fmt.Errorf("bert: scorer is required")
	}
	if cfg.NumLabels <= 0 {
		cfg.NumLabels = 3
	}
	return &BERTRouter{
		tokenizer: cfg.Tokenizer,
		scorer:    cfg.Scorer,
		numLabels: cfg.NumLabels,
	}, nil
}

// Score tokenizes the prompt, calls the Triton Scorer for logits,
// applies softmax, and returns the strong model's win rate.
//
// The model outputs logits for [strong_wins, tie, weak_wins].
// Win rate = 1 - P(tie) - P(weak_wins) = P(strong_wins).
// Following the Python implementation: binary_prob = softmax[-2:]
// so strong_win_rate = 1 - binary_prob.
func (r *BERTRouter) Score(
	ctx context.Context, prompt string,
) (float64, error) {
	tokenIDs, err := r.tokenizer.Tokenize(prompt)
	if err != nil {
		return 0, fmt.Errorf("bert: tokenize: %w", err)
	}

	// Convert int32 token IDs to float32 for Triton.
	input := make([]float32, len(tokenIDs))
	for i, id := range tokenIDs {
		input[i] = float32(id)
	}

	// Call Triton — expects logits back via the Score method.
	// For BERT, we get a single score from Triton; the full logits
	// would require a custom inference protocol. Here we interpret
	// the scalar output as the win rate directly when Triton
	// handles softmax internally.
	//
	// When Triton returns raw logits, use LogitsScore instead.
	winRate, err := r.scorer.Score(ctx, input)
	if err != nil {
		return 0, fmt.Errorf("bert: scorer: %w", err)
	}

	return winRate, nil
}

// ScoreFromLogits computes the strong model win rate from raw logits.
// This implements the Python BERTRouter softmax logic:
//
//	softmax(logits) -> binary_prob = sum(softmax[-2:])
//	strong_win_rate = 1 - binary_prob
func ScoreFromLogits(logits []float64) float64 {
	if len(logits) == 0 {
		return 0.5
	}

	softmax := Softmax(logits)

	// binary_prob = sum of last 2 classes (tie + weak wins).
	var binaryProb float64
	start := len(softmax) - 2
	if start < 0 {
		start = 0
	}
	for i := start; i < len(softmax); i++ {
		binaryProb += softmax[i]
	}

	return 1.0 - binaryProb
}

// Softmax computes the softmax of a slice of logits.
func Softmax(logits []float64) []float64 {
	if len(logits) == 0 {
		return nil
	}

	// Numerical stability: subtract max.
	maxVal := logits[0]
	for _, v := range logits[1:] {
		if v > maxVal {
			maxVal = v
		}
	}

	expScores := make([]float64, len(logits))
	var sum float64
	for i, v := range logits {
		expScores[i] = math.Exp(v - maxVal)
		sum += expScores[i]
	}

	result := make([]float64, len(logits))
	for i := range expScores {
		result[i] = expScores[i] / sum
	}
	return result
}
